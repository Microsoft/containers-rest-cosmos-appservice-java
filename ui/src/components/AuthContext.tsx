import { createContext } from 'react'

export interface IAuthContextValues {
  accessToken: string,
  authResponse: string,
  handleAuth: () => void,
  setAuthResponse: (msg?: string) => void,
}

// Default values and enforce use of interface
const defaultAuthContextValue: IAuthContextValues = {
  accessToken: null,
  authResponse: null,
  handleAuth: () => null,
  setAuthResponse: () => null,
}

export const AuthContext = createContext<IAuthContextValues>(defaultAuthContextValue)
// exporting the Provider and Consumer components for more specific imports
export const AuthProvider = AuthContext.Provider
export const AuthConsumer = AuthContext.Consumer
