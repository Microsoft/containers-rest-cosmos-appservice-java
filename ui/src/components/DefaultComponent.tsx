import * as React from 'react'

export interface IDefaultComponentProps {
    default: boolean,
}

export interface IDefaultComponentState {
    deerSrc: string,
}

export class DefaultComponent extends React.Component<IDefaultComponentProps> {
    public state = {
        deerSrc: null,
    }

    public componentWillMount() {
        this.setState({ deerSrc: 'https://i.gifer.com/KG8.gif' })
    }

    public render() {
        return (
            <div className='default-component cover'>
                <h1>
                    Oh Deer :(
                <br />
                    <small>
                        Error 404
                    </small>
                </h1>
                <p className='lead'>
                    The requested resource could not be found but may be available again in the future.
            </p>
                <img className='default-page-image' src={this.state.deerSrc} />
            </div>
        )
    }
}
