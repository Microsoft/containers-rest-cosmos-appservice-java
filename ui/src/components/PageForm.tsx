import * as React from 'react'

export interface IPageFormProps {
  inputTitle: string,
  inputPlaceholder: string,
  onInputChange: (event) => void,
  onSubmitClick: (event) => void,
}

export function PageForm(props: IPageFormProps) {
  return (
    <form className='form'>
      <label>
        <span className='form-field-pre-text'>{props.inputTitle}</span>
        <input
          className='form-field-input'
          type='text'
          name='id'
          onChange={props.onInputChange}
          placeholder={props.inputPlaceholder}
        />
        <span className='form-field-sub-text'>Leave empty for random sample</span>
      </label>
      <input
        className='form-field-submit'
        type='submit'
        value='Search'
        onClick={props.onSubmitClick}
      />
    </form>
  )
}
