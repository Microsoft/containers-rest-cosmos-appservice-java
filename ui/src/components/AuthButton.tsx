import * as React from 'react'

import { AuthContext } from './AuthContext'

export class AuthButton extends React.Component {
    public static contextType = AuthContext
    public render() {
        return (
            <button
                onClick={this.context.handleAuth}
                className='login-button'
            >
                {this.context.accessToken ? 'Log Out' : 'Log In'}
            </button>
        )
    }
}
