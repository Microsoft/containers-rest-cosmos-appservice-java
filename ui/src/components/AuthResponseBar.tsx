import * as React from 'react'

import { AuthContext } from './AuthContext'

export class AuthResponseBar extends React.Component {
  public static contextType = AuthContext
  public render() {
    return this.context.authResponse && (
      <div className='auth-message-container'>
        <button className='auth-message-close-button' onClick={() => this.context.setAuthResponse(null)}>X</button>
        <span className='auth-message'>{this.context.authResponse}</span>
      </div>
    )
  }
}
