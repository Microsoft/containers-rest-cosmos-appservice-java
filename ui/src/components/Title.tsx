import * as React from 'react'

import { AuthContext } from './AuthContext'
import { PageForm } from './PageForm'

export interface ITitleProps { path: string }

export interface ITitleState { loading: boolean, titleId: string, result: object }

export interface ITitleResponse extends Response {
  _embedded?: { titles: Array<{ tconst: string }> },
  error?: string,
  error_description?: string
}

export interface ITitleResult { tconst: string }

export class Title extends React.Component<ITitleProps, ITitleState> {
  public static contextType = AuthContext

  public state = {
    loading: false,
    result: null,
    titleId: null,
  }

  public render() {
    return (
      <div className='page-container'>
        <div className='form-container'>
          <h1>Search Titles</h1>
          <p className='form-description'>Enter the ID for a title to get their information from the database.
            Leave the form empty for a random sample of people from the database.</p>
          <PageForm
            inputTitle='Title ID:'
            inputPlaceholder='Title ID'
            onInputChange={this.handleNameInputChange}
            onSubmitClick={this.handleFormSubmit}
          />
        </div>
        <div className='results-container'>
          <h2 className='results-title'>{`Results for TitleId: ${this.state.result ? this.state.titleId : ''}`}</h2>
          <pre className='results-view'>{JSON.stringify(this.state.result, null, 2)}</pre>
          <h4>{this.state.loading ? 'Loading. . .' : null}</h4>
        </div>
      </div>
    )
  }

  private handleNameInputChange = (event) => this.setState({
    result: null,
    titleId: event.target.value,
  })

  private handleFormSubmit = async (event) => {
    event.preventDefault()

    this.setState({ loading: true })

    // set up endpoint
    const id = this.state.titleId && this.state.titleId.replace(/\s+/g, '')
    const base = WEBPACK_PROP_TITLE_BASE_URL
    const endpoint = id ? base + id : base

    // set up request header with Bearer token
    const headers = new Headers()
    const bearer = `Bearer ${this.context.accessToken}`
    headers.append('Authorization', bearer)
    const options = {
      headers,
      method: 'GET',
    }

    try {
      const response = await fetch(endpoint, options)
      const resOut: ITitleResponse = await response.json()

      if (resOut.hasOwnProperty('error')) {
        throw new Error(`Error: ${resOut.error}, ${resOut.error_description}`)
      } else if (id) {
        this.setState({ result: resOut })
      } else {
        const titles = resOut._embedded.titles
        const result: ITitleResult = titles[Math.floor(Math.random() * titles.length)]
        this.setState({ result, titleId: result.tconst })
      }
    } catch (err) {
      this.setState({ result: { error: err.message } })
    } finally {
      this.setState({ loading: false })
    }
  }
}
