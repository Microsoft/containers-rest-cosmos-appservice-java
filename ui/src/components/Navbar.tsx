import { Link } from '@reach/router'
import * as React from 'react'

import { AuthButton } from './AuthButton'

import './../styles/Navbar.css'

const isActive = ({ isCurrent }) => {
    return {
        className: isCurrent ? 'nav-link active' : 'nav-link',
    }
}

export class Navbar extends React.Component {
    public render() {
        return (
            <nav className='nav-container'>
                <span className='nav-title'>
                    Project Jackson
                </span>
                <div className='nav-link-container'>
                    <Link getProps={isActive} to='/'>Home</Link>
                    <Link getProps={isActive} to='/people'>People</Link>
                    <Link getProps={isActive} to='/titles'>Titles</Link>
                    <AuthButton />
                </div>
            </nav>
        )
    }
}
