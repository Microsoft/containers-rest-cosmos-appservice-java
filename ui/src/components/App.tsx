import { Router } from '@reach/router'
import * as Msal from 'msal'
import * as React from 'react'

import { AuthProvider } from './AuthContext'
import { AuthResponseBar } from './AuthResponseBar'
import { DefaultComponent } from './DefaultComponent'
import { Home } from './Home'
import { Navbar } from './Navbar'
import { Person } from './Person'
import { PrivateRoute } from './PrivateRoute'
import { Title } from './Title'

import './../styles/App.css'

const appConfig = {
    clientID: WEBPACK_PROP_AAD_CLIENT_ID,
}

// initialize the UserAgentApplication globally so popup and iframe can run in the background
const userAgentApp = appConfig.clientID && new Msal.UserAgentApplication(appConfig.clientID, null, null)
// short circuit userAgentApp. If clientID is an empty string so is userAgentApp

export class App extends React.Component {

    // App is still responsible for managing the auth state
    // it uses the AuthContext to share this with other aspects of the UI
    // including the navbar, private route component, and the pages themselves
    public state = {
        accessToken: null,
        authResponse: null,
    }

    public handleAuth = async () => {
        if (!appConfig.clientID) {
            if (this.state.accessToken !== null) {
                this.setState({ accessToken: null })
            } else {
                // tslint:disable-next-line no-console max-line-length
                console.warn('AAD Client ID has not been configured. If you are currently in production mode, see the \'deploy\' documentation for details on how to fix this.')
                this.setState({ accessToken: '' })
            }
        } else {
            let accessToken = null
            try {
                if (this.state.accessToken) {
                    // log out
                    await userAgentApp.logout()
                } else {
                    // log in
                    if (!appConfig.clientID) {
                        throw new Error(
                            'AAD Client ID has not been configured. See the \'deploy\' documentation for more details.')
                    }
                    const graphScopes = [appConfig.clientID]
                    await userAgentApp.loginPopup(graphScopes)
                    accessToken = await userAgentApp.acquireTokenSilent(graphScopes,
                        'https://login.microsoftonline.com/microsoft.onmicrosoft.com')
                }
                this.setState({ accessToken })
            } catch (err) {
                this.setAuthResponse(err.toString())
            }
        }
    }

    public setAuthResponse = (msg: string) => {
        this.setState({ authResponse: msg })
    }

    public render() {
        // Implementing the authprovider at app root to share the auth state
        // with all internal components (if they subscribe to it)
        // by linking the accessToken to the app state we can be certain the
        // context will always update and propogate the value to subscribed nodes
        return (
            <AuthProvider value={{
                accessToken: this.state.accessToken,
                authResponse: this.state.authResponse,
                handleAuth: this.handleAuth,
                setAuthResponse: this.setAuthResponse,
            }}>
                <div className='app-container'>
                    <Navbar />
                    <AuthResponseBar />
                    <Router>
                        <Home path='/' />
                        <PrivateRoute as={Person} path='/people' />
                        <PrivateRoute as={Title} path='/titles' />
                        <DefaultComponent default />
                    </Router>
                </div>
            </AuthProvider>
        )
    }
}
