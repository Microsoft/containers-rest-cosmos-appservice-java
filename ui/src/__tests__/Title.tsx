import * as React from 'react'
import * as renderer from 'react-test-renderer'

import { Title } from '../components/Title'

describe('<Title />', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(<Title path='/titles'/>)
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
