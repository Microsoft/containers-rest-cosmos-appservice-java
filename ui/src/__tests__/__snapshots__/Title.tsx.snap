// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Title /> renders correctly 1`] = `
<div
  className="page-container"
>
  <div
    className="form-container"
  >
    <h1>
      Search Titles
    </h1>
    <p
      className="form-description"
    >
      Enter the ID for a title to get their information from the database. Leave the form empty for a random sample of people from the database.
    </p>
    <form
      className="form"
    >
      <label>
        <span
          className="form-field-pre-text"
        >
          Title ID:
        </span>
        <input
          className="form-field-input"
          name="id"
          onChange={[Function]}
          placeholder="Title ID"
          type="text"
        />
        <span
          className="form-field-sub-text"
        >
          Leave empty for random sample
        </span>
      </label>
      <input
        className="form-field-submit"
        onClick={[Function]}
        type="submit"
        value="Search"
      />
    </form>
  </div>
  <div
    className="results-container"
  >
    <h2
      className="results-title"
    >
      Results for TitleId: 
    </h2>
    <pre
      className="results-view"
    >
      null
    </pre>
    <h4 />
  </div>
</div>
`;
